docs/math/number-theory/du.md

```diff
@@ -53,59 +53,99 @@ $$
     \sum_{k=1}^n (f*g)(k)=\mathrm{i}\frac{n(n+1)}{2}
     $$
     
-    计算 $\sum_k (f*g)(k)$ 和 $g$ 的时间复杂度均为 $O(1)$, 故可以考虑使用杜教筛。
+    计算 $\sum_k (f*g)(k)$ 和 $\sum_k g(k)$ 的时间复杂度均为 $O(1)$, 故可以考虑使用杜教筛。
 
 ## 时间复杂度
 
-我们认为计算 $\sum_{i=1}^n(f * g)(i)$ 和 $g(n)$ 的时间复杂度均为 $O(1)$, 设计算 $S(n)$ 的复杂度为 $T(n)$, 此时我们不妨将 $S(n)$ 简化为如下形式：
+令 $R(n)=\left\{\left\lfloor \dfrac{n}{k} \right\rfloor: k=2,3,\dots,n\right\}$, 我们有如下引理：
 
-$$
-S(n)=\sum_{i=2}^n S\left(\left\lfloor\frac{n}{i}\right\rfloor\right)
-$$
-
-由 [整除分块/数论分块](./sqrt-decomposition.md) 可知 $\left\lfloor \dfrac n i \right\rfloor$ 共有 $O(\sqrt{n})$ 种取值，故有：
-
-$$
-T(n)=O\left(\sqrt{n}\right)+O\left(\sum_{i=2}^{\sqrt{n}} T\left(\left\lfloor\frac{n}{i}\right\rfloor\right)\right)
-$$
-
-$$
-\begin{aligned}
-    T\left(\left\lfloor\frac{n}{i}\right\rfloor\right) & =O\left(\sqrt{\frac{n}{i}}\right)+O\left(\sum_{j=2}^{\sqrt{\frac{n}{i}}} T\left(\left\lfloor\frac{n}{ij}\right\rfloor\right)\right) \\
-                                                       & =O\left(\sqrt{\frac{n}{i}}\right)
-\end{aligned}
-$$
+???+ abstract "引理"
+    对任意的 $m\in R(n)$，我们有 $R(m)\subseteq R(n)$.
+    
+    ???+ note "证明"
+        令 $m=\left\lfloor\dfrac{n}{x}\right\rfloor$, 任取 $\left\lfloor\dfrac{m}{y}\right\rfloor \in R(m)$, 由整除分块/数论分块的 [引理 1](./sqrt-decomposition.md#引理-1) 可知：
+        
+        $$
+        \left\lfloor\dfrac{m}{y}\right\rfloor=\left\lfloor\dfrac{n}{xy}\right\rfloor\in R(n).
+        $$
 
-???+ note
-    $O\left(\sum_{j=2}^{\sqrt{\frac{n}{i}}} T\left(\left\lfloor\frac{n}{ij}\right\rfloor\right)\right)$ 视作高阶无穷小，从而可以舍去。
+设计算 $\sum_{i=1}^n(f * g)(i)$ 和 $\sum_{i=1}^n g(i)$ 的时间复杂度均为 $O(1)$. 由引理可知：使用记忆化之后，每个 $S(k)~(k\in R(n))$ 均只会计算一次。
 
-故：
+由整除分块/数论分块的 [引理 2](./sqrt-decomposition.md) 可知 $|R(n)|=2\sqrt{n}+\Theta(1)$. 设计算 $S(n)$ 的时间复杂度为 $T(n)$, 则：
 
 $$
 \begin{aligned}
-    T(n) & =O\left(\sqrt{n}\right)+O\left(\sum_{i=2}^{\sqrt{n}} \sqrt{\frac{n}{i}}\right)=O\left(\sum_{i=2}^{\sqrt{n}} \sqrt{\frac{n}{i}}\right) \\
-         & =O\left(\int_{0}^{\sqrt{n}}\sqrt{\frac{n}{x}}\mathrm{d}x\right)                                                                       \\
-         & =O\left(n^{\frac{3}{4}}\right)
+    T(n) & = \sum_{k\in R(n)} T(k)\\
+         & = \Theta(\sqrt n)+\sum_{k=1}^{\lfloor\sqrt n\rfloor} O(\sqrt k)+\sum_{k=2}^{\lfloor\sqrt n\rfloor} O\left(\sqrt{\dfrac{n}{k}}\right)\\
+         & = O\left(\int_{0}^{\sqrt n} \sqrt{x} + \sqrt{\dfrac{n}{x}} \mathrm{d}x\right)\\
```

Comment:
Is parentheses needed here? 🤔 It has been several years since I wrote my last integral symbol.

Reply:
> Would adding a parentheses here makes it better? 🤔 Not sure (there has been several years since I wrote my last integral symbol).

In my opinion, $\int$ and $\mathrm{d}x$ are just like a pair of brackets, but maybe it will cause some misunderstanding, so I will add it.

Reply:
> Would adding a parentheses here makes it better? 🤔 Not sure (there has been several years since I wrote my last integral symbol).

In my opinion, $\int$ and $\mathrm{d}x$ are just like a pair of brackets, but maybe it will cause some misunderstanding, so I will add it.